// react-axios-provider-kit/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  AxiosContext: () => AxiosContext,
  AxiosProvider: () => AxiosProvider,
  createApi: () => createApi,
  createLocalStorageTokenStorage: () => createLocalStorageTokenStorage,
  createMemoryTokenStorage: () => createMemoryTokenStorage,
  createSessionStorageTokenStorage: () => createSessionStorageTokenStorage,
  useAxios: () => useAxios
});
module.exports = __toCommonJS(index_exports);

// src/AxiosProvider.tsx
var import_react2 = require("react");
var import_axios2 = __toESM(require("axios"), 1);

// src/context.ts
var import_react = require("react");
var AxiosContext = (0, import_react.createContext)(void 0);

// src/create-api.ts
var import_axios = __toESM(require("axios"), 1);
function createApi(client, opts) {
  const requestHandler = async (request) => {
    var _a, _b, _c, _d;
    try {
      const res = await request();
      return res.data;
    } catch (err) {
      if (import_axios.default.isAxiosError(err)) {
        const status = (_a = err.response) == null ? void 0 : _a.status;
        const url = (_c = (_b = err.config) == null ? void 0 : _b.url) == null ? void 0 : _c.toString();
        if ((opts == null ? void 0 : opts.onNotify) && !((_d = err.config) == null ? void 0 : _d._skipNotify)) {
          opts.onNotify({
            level: "error",
            message: "Request failed",
            status,
            url,
            code: err.code
          });
        }
      }
      throw err;
    }
  };
  return {
    get: (url, config) => requestHandler(() => client.get(url, config)),
    post: (url, data, config) => requestHandler(() => client.post(url, data, config)),
    put: (url, data, config) => requestHandler(() => client.put(url, data, config)),
    patch: (url, data, config) => requestHandler(() => client.patch(url, data, config)),
    delete: (url, config) => requestHandler(() => client.delete(url, config)),
    postBlob: async (url, data, config) => {
      const res = await client.post(url, data, { ...config, responseType: "blob" });
      return res.data;
    },
    uploadFile: (url, formData, config) => requestHandler(
      () => client.post(url, formData, {
        ...config,
        headers: {
          ...(config == null ? void 0 : config.headers) || {},
          "Content-Type": "multipart/form-data"
        }
      })
    )
  };
}

// src/AxiosProvider.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var defaultGetTimeZone = () => Intl.DateTimeFormat().resolvedOptions().timeZone;
var defaultBuildAuthHeader = (t) => `Bearer ${t}`;
var defaultCreateClient = ({ baseURL }) => import_axios2.default.create({ baseURL });
var defaultIsRefreshRequest = (config) => {
  var _a;
  return ((_a = config.url) != null ? _a : "").toString().includes("/auth/refresh");
};
function AxiosProvider({
  children,
  baseURL,
  tokenStorage,
  getTimeZone = defaultGetTimeZone,
  buildAuthHeader = defaultBuildAuthHeader,
  createMainClient = defaultCreateClient,
  createRefreshClient = ({ baseURL: baseURL2 }) => import_axios2.default.create({
    baseURL: baseURL2,
    withCredentials: true
  }),
  onAuthFailure,
  onNotify,
  shouldRefresh,
  shouldRetry,
  isRefreshRequest = defaultIsRefreshRequest,
  refreshAccessToken,
  exposeApi = true
}) {
  const storage = (0, import_react2.useMemo)(
    () => tokenStorage != null ? tokenStorage : {
      get: () => void 0,
      set: () => void 0
    },
    [tokenStorage]
  );
  const [tokenState, setTokenState] = (0, import_react2.useState)(() => storage.get());
  const tokenRef = (0, import_react2.useRef)(tokenState);
  (0, import_react2.useEffect)(() => {
    tokenRef.current = tokenState;
  }, [tokenState]);
  const setToken = (0, import_react2.useCallback)(
    (t) => {
      storage.set(t);
      tokenRef.current = t;
      setTokenState(t);
    },
    [storage]
  );
  const client = (0, import_react2.useMemo)(() => createMainClient({ baseURL }), [baseURL, createMainClient]);
  const refreshClient = (0, import_react2.useMemo)(
    () => createRefreshClient({ baseURL }),
    [baseURL, createRefreshClient]
  );
  const refreshPromiseRef = (0, import_react2.useRef)(null);
  const doNotify = (0, import_react2.useCallback)(
    (payload) => {
      onNotify == null ? void 0 : onNotify(payload);
    },
    [onNotify]
  );
  const defaultShouldRefresh = (0, import_react2.useCallback)((error, config) => {
    var _a;
    const status = (_a = error.response) == null ? void 0 : _a.status;
    return status === 401 && !config._retry;
  }, []);
  const shouldRefreshFn = shouldRefresh != null ? shouldRefresh : defaultShouldRefresh;
  const refreshToken = (0, import_react2.useCallback)(async () => {
    if (!refreshAccessToken) {
      throw new Error("refreshAccessToken is not configured.");
    }
    if (refreshPromiseRef.current) return refreshPromiseRef.current;
    const task = (async () => {
      var _a;
      const current = (_a = tokenRef.current) != null ? _a : storage.get();
      if (!current) throw new Error("Missing token.");
      const res = await refreshAccessToken({ refreshClient, token: current });
      if (!(res == null ? void 0 : res.accessToken)) throw new Error("Refresh returned empty token.");
      setToken(res.accessToken);
      return res.accessToken;
    })();
    refreshPromiseRef.current = task;
    try {
      return await task;
    } finally {
      refreshPromiseRef.current = null;
    }
  }, [refreshAccessToken, refreshClient, setToken, storage]);
  const failAuth = (0, import_react2.useCallback)(
    (reason) => {
      var _a;
      const lastToken = (_a = tokenRef.current) != null ? _a : storage.get();
      setToken(void 0);
      onAuthFailure == null ? void 0 : onAuthFailure({ error: reason.error, lastToken });
      doNotify({ level: "error", message: "Authentication required" });
    },
    [doNotify, onAuthFailure, setToken, storage]
  );
  (0, import_react2.useEffect)(() => {
    const reqId = client.interceptors.request.use(
      (config) => {
        var _a, _b;
        config.headers = (_a = config.headers) != null ? _a : {};
        config.headers["X-User-Time-Zone"] = getTimeZone();
        const t = (_b = tokenRef.current) != null ? _b : storage.get();
        if (t) config.headers.Authorization = buildAuthHeader(t);
        else delete config.headers.Authorization;
        return config;
      },
      (e) => Promise.reject(e)
    );
    const resId = client.interceptors.response.use(
      (r) => r,
      async (error) => {
        var _a, _b;
        const original = error.config;
        if (!original) return Promise.reject(error);
        if (error.code === "ERR_NETWORK") {
          if (!original._skipNotify) doNotify({
            level: "warning",
            message: "Network error",
            code: error.code
          });
          return Promise.reject(error);
        }
        if ((shouldRetry == null ? void 0 : shouldRetry(error, original)) && !original._retry) {
          original._retry = true;
          return client(original);
        }
        if (shouldRefreshFn(error, original)) {
          original._retry = true;
          if (isRefreshRequest(original)) {
            failAuth({ error });
            return Promise.reject(error);
          }
          try {
            const newToken = await refreshToken();
            original.headers = (_a = original.headers) != null ? _a : {};
            original.headers.Authorization = buildAuthHeader(newToken);
            return client(original);
          } catch (e) {
            failAuth({ error: e });
            return Promise.reject(e);
          }
        }
        if (((_b = error.response) == null ? void 0 : _b.status) === 403) {
          if (!original._skipNotify) doNotify({ level: "error", message: "Forbidden", status: 403 });
        }
        return Promise.reject(error);
      }
    );
    return () => {
      client.interceptors.request.eject(reqId);
      client.interceptors.response.eject(resId);
    };
  }, [
    client,
    buildAuthHeader,
    doNotify,
    failAuth,
    getTimeZone,
    isRefreshRequest,
    refreshToken,
    shouldRetry,
    shouldRefreshFn,
    storage
  ]);
  const api = (0, import_react2.useMemo)(() => exposeApi ? createApi(client, { onNotify }) : void 0, [client, exposeApi, onNotify]);
  const value = (0, import_react2.useMemo)(
    () => ({
      client,
      refreshClient,
      token: tokenState,
      setToken,
      api
    }),
    [client, refreshClient, tokenState, setToken, api]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(AxiosContext.Provider, { value, children });
}

// src/useAxios.ts
var import_react3 = require("react");
function useAxios() {
  const ctx = (0, import_react3.useContext)(AxiosContext);
  if (!ctx) throw new Error("useAxios must be used within an AxiosProvider");
  return ctx;
}

// src/storage.ts
function createSessionStorageTokenStorage(key) {
  return {
    get: () => {
      var _a;
      if (typeof window === "undefined") return void 0;
      return (_a = window.sessionStorage.getItem(key)) != null ? _a : void 0;
    },
    set: (token) => {
      if (typeof window === "undefined") return;
      if (token) window.sessionStorage.setItem(key, token);
      else window.sessionStorage.removeItem(key);
    }
  };
}
function createLocalStorageTokenStorage(key) {
  return {
    get: () => {
      var _a;
      if (typeof window === "undefined") return void 0;
      return (_a = window.localStorage.getItem(key)) != null ? _a : void 0;
    },
    set: (token) => {
      if (typeof window === "undefined") return;
      if (token) window.localStorage.setItem(key, token);
      else window.localStorage.removeItem(key);
    }
  };
}
function createMemoryTokenStorage(initial) {
  let token = initial;
  return {
    get: () => token,
    set: (t) => {
      token = t;
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AxiosContext,
  AxiosProvider,
  createApi,
  createLocalStorageTokenStorage,
  createMemoryTokenStorage,
  createSessionStorageTokenStorage,
  useAxios
});
