// react-axios-provider-kit/dist/index.js
// src/AxiosProvider.tsx
import { useCallback, useEffect, useMemo, useRef, useState } from "react";
import axios2 from "axios";

// src/context.ts
import { createContext } from "react";
var AxiosContext = createContext(void 0);

// src/create-api.ts
import axios from "axios";
function createApi(client, opts) {
  const requestHandler = async (request) => {
    var _a, _b, _c, _d;
    try {
      const res = await request();
      return res.data;
    } catch (err) {
      if (axios.isAxiosError(err)) {
        const status = (_a = err.response) == null ? void 0 : _a.status;
        const url = (_c = (_b = err.config) == null ? void 0 : _b.url) == null ? void 0 : _c.toString();
        if ((opts == null ? void 0 : opts.onNotify) && !((_d = err.config) == null ? void 0 : _d._skipNotify)) {
          opts.onNotify({
            level: "error",
            message: "Request failed",
            status,
            url,
            code: err.code
          });
        }
      }
      throw err;
    }
  };
  return {
    get: (url, config) => requestHandler(() => client.get(url, config)),
    post: (url, data, config) => requestHandler(() => client.post(url, data, config)),
    put: (url, data, config) => requestHandler(() => client.put(url, data, config)),
    patch: (url, data, config) => requestHandler(() => client.patch(url, data, config)),
    delete: (url, config) => requestHandler(() => client.delete(url, config)),
    postBlob: async (url, data, config) => {
      const res = await client.post(url, data, { ...config, responseType: "blob" });
      return res.data;
    },
    uploadFile: (url, formData, config) => requestHandler(
      () => client.post(url, formData, {
        ...config,
        headers: {
          ...(config == null ? void 0 : config.headers) || {},
          "Content-Type": "multipart/form-data"
        }
      })
    )
  };
}

// src/AxiosProvider.tsx
import { jsx } from "react/jsx-runtime";
var defaultGetTimeZone = () => Intl.DateTimeFormat().resolvedOptions().timeZone;
var defaultBuildAuthHeader = (t) => `Bearer ${t}`;
var defaultCreateClient = ({ baseURL }) => axios2.create({ baseURL });
var defaultIsRefreshRequest = (config) => {
  var _a;
  return ((_a = config.url) != null ? _a : "").toString().includes("/auth/refresh");
};
function AxiosProvider({
  children,
  baseURL,
  tokenStorage,
  getTimeZone = defaultGetTimeZone,
  buildAuthHeader = defaultBuildAuthHeader,
  createMainClient = defaultCreateClient,
  createRefreshClient = ({ baseURL: baseURL2 }) => axios2.create({
    baseURL: baseURL2,
    withCredentials: true
  }),
  onAuthFailure,
  onNotify,
  shouldRefresh,
  shouldRetry,
  isRefreshRequest = defaultIsRefreshRequest,
  refreshAccessToken,
  exposeApi = true
}) {
  const storage = useMemo(
    () => tokenStorage != null ? tokenStorage : {
      get: () => void 0,
      set: () => void 0
    },
    [tokenStorage]
  );
  const [tokenState, setTokenState] = useState(() => storage.get());
  const tokenRef = useRef(tokenState);
  useEffect(() => {
    tokenRef.current = tokenState;
  }, [tokenState]);
  const setToken = useCallback(
    (t) => {
      storage.set(t);
      tokenRef.current = t;
      setTokenState(t);
    },
    [storage]
  );
  const client = useMemo(() => createMainClient({ baseURL }), [baseURL, createMainClient]);
  const refreshClient = useMemo(
    () => createRefreshClient({ baseURL }),
    [baseURL, createRefreshClient]
  );
  const refreshPromiseRef = useRef(null);
  const doNotify = useCallback(
    (payload) => {
      onNotify == null ? void 0 : onNotify(payload);
    },
    [onNotify]
  );
  const defaultShouldRefresh = useCallback((error, config) => {
    var _a;
    const status = (_a = error.response) == null ? void 0 : _a.status;
    return status === 401 && !config._retry;
  }, []);
  const shouldRefreshFn = shouldRefresh != null ? shouldRefresh : defaultShouldRefresh;
  const refreshToken = useCallback(async () => {
    if (!refreshAccessToken) {
      throw new Error("refreshAccessToken is not configured.");
    }
    if (refreshPromiseRef.current) return refreshPromiseRef.current;
    const task = (async () => {
      var _a;
      const current = (_a = tokenRef.current) != null ? _a : storage.get();
      if (!current) throw new Error("Missing token.");
      const res = await refreshAccessToken({ refreshClient, token: current });
      if (!(res == null ? void 0 : res.accessToken)) throw new Error("Refresh returned empty token.");
      setToken(res.accessToken);
      return res.accessToken;
    })();
    refreshPromiseRef.current = task;
    try {
      return await task;
    } finally {
      refreshPromiseRef.current = null;
    }
  }, [refreshAccessToken, refreshClient, setToken, storage]);
  const failAuth = useCallback(
    (reason) => {
      var _a;
      const lastToken = (_a = tokenRef.current) != null ? _a : storage.get();
      setToken(void 0);
      onAuthFailure == null ? void 0 : onAuthFailure({ error: reason.error, lastToken });
      doNotify({ level: "error", message: "Authentication required" });
    },
    [doNotify, onAuthFailure, setToken, storage]
  );
  useEffect(() => {
    const reqId = client.interceptors.request.use(
      (config) => {
        var _a, _b;
        config.headers = (_a = config.headers) != null ? _a : {};
        config.headers["X-User-Time-Zone"] = getTimeZone();
        const t = (_b = tokenRef.current) != null ? _b : storage.get();
        if (t) config.headers.Authorization = buildAuthHeader(t);
        else delete config.headers.Authorization;
        return config;
      },
      (e) => Promise.reject(e)
    );
    const resId = client.interceptors.response.use(
      (r) => r,
      async (error) => {
        var _a, _b;
        const original = error.config;
        if (!original) return Promise.reject(error);
        if (error.code === "ERR_NETWORK") {
          if (!original._skipNotify) doNotify({
            level: "warning",
            message: "Network error",
            code: error.code
          });
          return Promise.reject(error);
        }
        if ((shouldRetry == null ? void 0 : shouldRetry(error, original)) && !original._retry) {
          original._retry = true;
          return client(original);
        }
        if (shouldRefreshFn(error, original)) {
          original._retry = true;
          if (isRefreshRequest(original)) {
            failAuth({ error });
            return Promise.reject(error);
          }
          try {
            const newToken = await refreshToken();
            original.headers = (_a = original.headers) != null ? _a : {};
            original.headers.Authorization = buildAuthHeader(newToken);
            return client(original);
          } catch (e) {
            failAuth({ error: e });
            return Promise.reject(e);
          }
        }
        if (((_b = error.response) == null ? void 0 : _b.status) === 403) {
          if (!original._skipNotify) doNotify({ level: "error", message: "Forbidden", status: 403 });
        }
        return Promise.reject(error);
      }
    );
    return () => {
      client.interceptors.request.eject(reqId);
      client.interceptors.response.eject(resId);
    };
  }, [
    client,
    buildAuthHeader,
    doNotify,
    failAuth,
    getTimeZone,
    isRefreshRequest,
    refreshToken,
    shouldRetry,
    shouldRefreshFn,
    storage
  ]);
  const api = useMemo(() => exposeApi ? createApi(client, { onNotify }) : void 0, [client, exposeApi, onNotify]);
  const value = useMemo(
    () => ({
      client,
      refreshClient,
      token: tokenState,
      setToken,
      api
    }),
    [client, refreshClient, tokenState, setToken, api]
  );
  return /* @__PURE__ */ jsx(AxiosContext.Provider, { value, children });
}

// src/useAxios.ts
import { useContext } from "react";
function useAxios() {
  const ctx = useContext(AxiosContext);
  if (!ctx) throw new Error("useAxios must be used within an AxiosProvider");
  return ctx;
}

// src/storage.ts
function createSessionStorageTokenStorage(key) {
  return {
    get: () => {
      var _a;
      if (typeof window === "undefined") return void 0;
      return (_a = window.sessionStorage.getItem(key)) != null ? _a : void 0;
    },
    set: (token) => {
      if (typeof window === "undefined") return;
      if (token) window.sessionStorage.setItem(key, token);
      else window.sessionStorage.removeItem(key);
    }
  };
}
function createLocalStorageTokenStorage(key) {
  return {
    get: () => {
      var _a;
      if (typeof window === "undefined") return void 0;
      return (_a = window.localStorage.getItem(key)) != null ? _a : void 0;
    },
    set: (token) => {
      if (typeof window === "undefined") return;
      if (token) window.localStorage.setItem(key, token);
      else window.localStorage.removeItem(key);
    }
  };
}
function createMemoryTokenStorage(initial) {
  let token = initial;
  return {
    get: () => token,
    set: (t) => {
      token = t;
    }
  };
}
export {
  AxiosContext,
  AxiosProvider,
  createApi,
  createLocalStorageTokenStorage,
  createMemoryTokenStorage,
  createSessionStorageTokenStorage,
  useAxios
};
